compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/model/types/BFiniteType.java

```diff
@@ -65,6 +64,11 @@ public String toString() {
         return joiner.toString();
     }
 
+    @Override
+    public boolean hasImplicitInitialValue() {
+        return valueSpace.stream().anyMatch(value -> value.type.tag == TypeTags.NIL);
```

Comment:
For finite types, checking for just nil is not enough. As per the spec, the following finite type's implicit initial value would be 0.
```ballerina
type STATE 0|1|2;
```
Here's what the spec says regarding implicit initial values of union types:
```
The rules for the implicit initial value of a union type are as follows:
● if () is in the value space of a union type, then the implicit initial value for the union
type is ();
● otherwise if the value space only contains only values of a single basic type, and one
of those values is the implicit initial value of that basic type, then the implicit initial
value for the union is the implicit initial value of the basic type of its values (for
example, the implicit initial value for 0|1 is 0)
● otherwise the union type does not have an implicit initial value.
```